x-pack/plugin/repository-encrypted/src/main/java/org/elasticsearch/repositories/encrypted/EncryptedRepository.java

```diff
@@ -3,14 +3,847 @@
+                        /* fail in the exceptional case of metadata blob name conflict */);
+            }
+            // afterwards write the encrypted data blob
+            // prepended to the encrypted data blob is the unique identifier (fixed length) of the metadata blob
+            final long encryptedBlobSize = (long) MetadataIdentifier.byteLength() +
+                    EncryptionPacketsInputStream.getEncryptionLength(blobSize, PACKET_LENGTH_IN_BYTES);
+            try (InputStream encryptedInputStream =
+                         ChainingInputStream.chain(new ByteArrayInputStream(metadataIdentifier.asByteArray()),
+                                 new EncryptionPacketsInputStream(inputStream, dataEncryptionKey, nonce, PACKET_LENGTH_IN_BYTES))) {
+                delegatedBlobContainer.writeBlob(blobName, encryptedInputStream, encryptedBlobSize, failIfAlreadyExists);
+            }
+        }
+
+        @Override
+        public void writeBlobAtomic(String blobName, InputStream inputStream, long blobSize, boolean failIfAlreadyExists)
+                throws IOException {
+            // the encrypted repository does not offer an alternative implementation for atomic writes
+            // fallback to regular write
+            writeBlob(blobName, inputStream, blobSize, failIfAlreadyExists);
+        }
+
+        @Override
+        public DeleteResult delete() throws IOException {
+            // first delete the encrypted data blob
+            DeleteResult deleteResult = delegatedBlobContainer.delete();
+            // then delete metadata
+            try {
+                deleteResult = deleteResult.add(encryptionMetadataBlobContainer.delete());
+            } catch (IOException e) {
+                // the encryption metadata blob container might not exist at all
+                logger.warn("Failure to delete metadata blob container " + encryptionMetadataBlobContainer.path(), e);
+            }
+            return deleteResult;
+        }
+
+        @Override
+        public void deleteBlobsIgnoringIfNotExists(List<String> blobNames) throws IOException {
+            Objects.requireNonNull(blobNames);
+
+            // find all the blob names that must be deleted
+            Set<String> blobNamesSet = new HashSet<>(blobNames);
+            Set<String> blobNamesToDelete = new HashSet<>();
+            for (String existingBlobName : delegatedBlobContainer.listBlobs().keySet()) {
+                if (blobNamesSet.contains(existingBlobName)) {
+                    blobNamesToDelete.add(existingBlobName);
+                }
+            }
+
+            // find all the metadata blob names that must be deleted
+            Map<String, List<String>> blobNamesToMetadataNamesToDelete = new HashMap<>(blobNamesToDelete.size());
+            Set<String> allMetadataBlobNames = new HashSet<>();
+            try {
+                allMetadataBlobNames = encryptionMetadataBlobContainer.listBlobs().keySet();
```

Comment:
This could become an issue as it makes deleting very expensive for large paths.
I think we must find a cleaner way of dealing with this. One solution I see would be to not use the metadata naming including a UUID for all blobs. We could just do it for `snap-` and `meta-` blobs (these are the only ones that can technically get overwritten). I know it's another special case, but this would allow us to simply delete all other blobs' meta without listing the metadata path. For the `snap-` and `meta-` blobs we can then do a listing including the prefix in the meta directory and it's effectively always a small listing.

WDYT?